src/plugins/formulas/formulas.js

```diff
@@ -126,318 +162,427 @@ export class Formulas extends BasePlugin {
+   * @param {number} row Physical row height.
+   * @param {number} column Physical column index.
+   * @param {object} valueHolder Object which contains original value which can be modified by overwriting `.value`
+   *   property.
+   * @param {string} ioMode String which indicates for what operation hook is fired (`get` or `set`).
    */
   onModifyData(row, column, valueHolder, ioMode) {
-    if (ioMode === 'get' && this.hasComputedCellValue(row, column)) {
-      valueHolder.value = this.getCellValue(row, column);
+    if (!this.enabled || this.#internalOperationPending) {
+      // TODO check if this line is actually ever reached
+      return;
+    }
 
-    } else if (ioMode === 'set' && isFormulaExpression(valueHolder.value)) {
-      valueHolder.value = toUpperCaseFormula(valueHolder.value);
+    const address = {
+      row: this.hot.toVisualRow(row),
+      col: column,
+      sheet: this.sheetId
+    };
+
+    if (ioMode === 'get') {
+      const cellValue = this.engine.getCellValue(address);
+
+      // If `cellValue` is an object it is expected to be an error
+      const value = (typeof cellValue === 'object' && cellValue !== null) ? cellValue.value : cellValue;
+
+      valueHolder.value = value;
+    } else {
+      if (
+        !this.engine.isItPossibleToSetCellContents(address)
+      ) {
+        warn(`Not possible to set cell data at ${JSON.stringify(address)}`);
+
+        return;
+      }
+
+      this.#shouldSuspendRenders = true;
+      this.engine.setCellContents(address, valueHolder.value);
+      this.#shouldSuspendRenders = false;
     }
   }
 
   /**
-   * On before value render listener.
+   * `modifySourceData` hook callback.
    *
    * @private
-   * @param {*} value Value to render.
-   * @returns {*}
+   * @param {number} row Physical row index.
+   * @param {number} col Physical column index.
+   * @param {object} valueHolder Object which contains original value which can be modified by overwriting `.value`
+   *   property.
+   * @param {string} ioMode String which indicates for what operation hook is fired (`get` or `set`).
    */
-  onBeforeValueRender(value) {
-    let renderValue = value;
+  onModifySourceData(row, col, valueHolder, ioMode) {
+    if (!this.enabled || this.#internalOperationPending) {
+      return;
+    }
+
+    const dimensions = this.engine.getSheetDimensions(this.engine.getSheetId(this.sheetName));
 
-    if (isFormulaExpressionEscaped(renderValue)) {
-      renderValue = unescapeFormulaExpression(renderValue);
+    // Don't actually change the source data if HyperFormula is not
+    // initialized yet. This is done to allow the `afterLoadData` hook to
+    // load the existing source data with `Handsontable#getSourceDataArray`
+    // properly.
+    if (dimensions.width === 0 && dimensions.height === 0) {
+      return;
     }
 
-    return renderValue;
+    const address = {
+      row: this.hot.toVisualRow(row),
+      col: this.hot.propToCol(col),
```

Comment:
Missed this one. Shouldn't we use physical indexes with HF?

Reply:
https://github.com/handsontable/handsontable/pull/7648/commits/f95eacec468c5f2c33385a390be81c383c10506e